crates/shared/src/sources/balancer_v2/pool_fetching/registry.rs

```diff
@@ -0,0 +1,174 @@
+    Web3, Web3CallBatch, Web3Transport,
+};
+use anyhow::Result;
+use contracts::{balancer_v2_base_pool_factory, BalancerV2BasePoolFactory};
+use ethcontract::{errors::MethodError, BlockId, Instance, H256};
+use futures::future;
+use model::TokenPair;
+use std::{collections::HashSet, sync::Arc};
+use tokio::sync::Mutex;
+
+use super::internal::InternalPoolFetching;
+
+impl_event_retrieving! {
+    pub BasePoolFactoryContract for balancer_v2_base_pool_factory
+}
+
+/// Type alias for the internal event updater type.
+type PoolUpdater<Factory> =
+    Mutex<EventHandler<Web3, BasePoolFactoryContract, PoolStorage<Factory>>>;
+
+/// The Pool Registry maintains an event handler for each of the Balancer Pool Factory contracts
+/// and maintains a `PoolStorage` for each.
+/// Pools are read from this registry, via the public method `pool_ids_for_token_pairs`
+/// which takes a collection of `TokenPair`, gets the relevant pools from each `PoolStorage`
+/// and returns a merged de-duplicated version of the results.
+pub struct Registry<Factory>
+where
+    Factory: FactoryIndexing,
+{
+    web3: Web3,
+    fetcher: Arc<dyn PoolInfoFetching<Factory>>,
+    updater: PoolUpdater<Factory>,
+}
+
+impl<Factory> Registry<Factory>
+where
+    Factory: FactoryIndexing,
+{
+    /// Returns a new pool registry for the specified factory.
+    pub fn new(
+        fetcher: Arc<dyn PoolInfoFetching<Factory>>,
+        factory_instance: &Instance<Web3Transport>,
+        initial_pools: Vec<Factory::PoolInfo>,
+        start_sync_at_block: Option<u64>,
+    ) -> Self {
+        let web3 = factory_instance.web3();
+        let updater = Mutex::new(EventHandler::new(
+            web3.clone(),
+            BasePoolFactoryContract(base_pool_factory(factory_instance)),
+            PoolStorage::new(initial_pools, fetcher.clone()),
+            start_sync_at_block,
+        ));
+        Self {
+            web3,
+            fetcher,
+            updater,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl<Factory> InternalPoolFetching for Registry<Factory>
+where
+    Factory: FactoryIndexing,
+{
+    async fn pool_ids_for_token_pairs(&self, token_pairs: HashSet<TokenPair>) -> HashSet<H256> {
+        self.updater
+            .lock()
+            .await
+            .store()
+            .pool_ids_for_token_pairs(&token_pairs)
+    }
+
+    async fn pools_by_id(&self, pool_ids: HashSet<H256>, block: Block) -> Result<Vec<Pool>> {
+        let mut batch = Web3CallBatch::new(self.web3.transport().clone());
+        let block = BlockId::Number(block.into());
+
+        let pool_infos = self.updater.lock().await.store().pools_by_id(&pool_ids);
+        let pool_futures = pool_infos
+            .into_iter()
+            .map(|pool_info| self.fetcher.fetch_pool(&pool_info, &mut batch, block))
+            .collect::<Vec<_>>();
+
+        batch.execute_all(MAX_BATCH_SIZE).await;
+
+        let pools = future::join_all(pool_futures).await;
+        collect_pool_results(pools)
```

Comment:
Also moved from `pool_cache`. This is just a generic version of the `CacheFetching<H256, WeightedPool>` and `CacheFetching<H256, StablePool>` implementations.

So less code duplication 🎉.

Reply:
ESL 😛